runtime/gc_base/ObjectAccessBarrier.cpp

```diff
@@ -1441,6 +1441,75 @@ MM_ObjectAccessBarrier::cloneIndexableObject(J9VMThread *vmThread, J9IndexableOb
 	return;
 }
 
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+/**
+ * Copy all of the fields of a value class instance to another value class instance.
+ * The source or destination may be a flattened value within another object, meaning
+ * srcOffset and destOffset need not be equal.
+ * @TODO This does not currently check if the fields that it is reading are volatile.
+ *
+ * @oaram valueClass The value class.
+ * @param srcObject The object containing the value class instance fields being copied.
+ * @param srcOffset The offset of the value class instance fields in srcObject.
+ * @param destValue The object containing the value class instance fields being copied to.
+ * @param destOffset The offset of the value class instance fields in destObject.
+ */
+void
+MM_ObjectAccessBarrier::copyValue(
+							J9VMThread *vmThread, J9Class *valueClass,
+							J9Object *srcObject, UDATA srcOffset,
+							J9Object *destObject, UDATA destOffset
+) {
+	/* subtract size of lockword */
+	const UDATA fieldsSize = _extensions->mixedObjectModel.getSizeInBytesWithoutHeader(valueClass) - sizeof(j9objectmonitor_t);
+	if (srcOffset & (sizeof(fj9object_t) - 1) || destOffset & (sizeof(fj9object_t) - 1)) {
+		/* Check whether the source or destination is not sizeof(ftj9object_t)-aligned.
```

Comment:
This will skip all of the object access read/write barriers.  I am not sure if that is correct in this context.  You may need to just have the code in the else statement to be correct.  Is your statement about no object references true for compressed references?